File: script/make/protocol/js/DailyProtocol.js
export default class DailyProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 12301: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 12302: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 12303: {
                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 日常ID
                view.setUint32(offset, data["dailyId"], false);
                offset = offset + 4;
                return new DataView(view.buffer.slice(0, offset));
            }
            case 12304: {
                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 阶段ID
                view.setUint32(offset, data["stageId"], false);
                offset = offset + 4;
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 12301: {
                // 统计列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Count
                    // 统计类型
                    const type = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 今日数量
                    const todayNumber = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const count = {type, todayNumber};
                    // add
                    list.push(count);
                }
                return {list};
            }
            case 12302: {
                // 日常列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Daily
                    // 日常ID
                    const dailyId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 是否领取奖励
                    const isAward = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const daily = {dailyId, isAward};
                    // add
                    list.push(daily);
                }
                // DailyActive
                // 奖励阶段ID
                const stageId = view.getUint32(offset, false);
                offset = offset + 4;
                // 活跃度
                const score = view.getUint32(offset, false);
                offset = offset + 4;
                // object
                const dailyActive = {stageId, score};
                return {list, dailyActive};
            }
            case 12303: {
                // 结果
                const resultLength = view.getUint16(offset, false);
                offset = offset + 2;
                const resultArray = new Uint8Array(view.buffer.slice(offset, offset + resultLength));
                const result = textDecoder.decode(resultArray);
                offset = offset + resultLength;
                return {result};
            }
            case 12304: {
                // 结果
                const resultLength = view.getUint16(offset, false);
                offset = offset + 2;
                const resultArray = new Uint8Array(view.buffer.slice(offset, offset + resultLength));
                const result = textDecoder.decode(resultArray);
                offset = offset + resultLength;
                return {result};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/lua/DailyProtocol.lua
DailyProtocol = {}

function DailyProtocol.encode(offset, protocol, data)
    if protocol == 12301 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 12302 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 12303 then
        local offset = offset
        local table = {}
        -- 日常ID
        table[offset] = string.pack(">I4", data["dailyId"])
        offset = offset + 1
        return table
    elseif protocol == 12304 then
        local offset = offset
        local table = {}
        -- 阶段ID
        table[offset] = string.pack(">I4", data["stageId"])
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function DailyProtocol.decode(offset, protocol, data)
    if protocol == 12301 then
        local offset = offset
        -- 统计列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Count
            -- 统计类型
            local type = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 今日数量
            local todayNumber = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- object
            local count = {type = type, todayNumber = todayNumber}
            list[listIndex] = count
        end
        return {list = list}
    elseif protocol == 12302 then
        local offset = offset
        -- 日常列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Daily
            -- 日常ID
            local dailyId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 是否领取奖励
            local isAward = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local daily = {dailyId = dailyId, isAward = isAward}
            list[listIndex] = daily
        end
        -- DailyActive
        -- 奖励阶段ID
        local stageId = string.unpack(">I4", data, offset)
        offset = offset + 4
        -- 活跃度
        local score = string.unpack(">I4", data, offset)
        offset = offset + 4
        -- object
        local dailyActive = {stageId = stageId, score = score}
        return {list = list, dailyActive = dailyActive}
    elseif protocol == 12303 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    elseif protocol == 12304 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end